### ISSUE TITLE
Automate deploy process

### ISSUE THREAD

**Issue by danfinlay** [MetaMask/metamask-extension#3389]
This issue represents the dream that someday we might be able to automatically deploy to all our build targets & platforms without visiting a bunch of websites.

I'll be marking this as an epic in ZenHub, blocked by various strategies.

**Comment by lazaridiscom** [comment]
"epic in ZenHub" - "blocked by various strategies", what does this mean (standard technical English please) ?

**Comment by kumavis** [comment]
@lazaridiscom  https://app.zenhub.com

**Comment by danfinlay** [comment]
Yeah, I was thinking "deploy strategies", but yeah was bad wording. Updated top post.

**Comment by lazaridiscom** [comment]
... ... ... ^%*&%*^@#% - RUUUN!!!

(My experience says: "Whenever you see "Agile" within a creative project.... RUN!!!")

**Comment by lazaridiscom** [comment]
@danfinlay , some comments:

> This is an "epic" in ZenHub, i.e. it has many "dependencies"/blockers, that can be easily viewed in ZenHub.

If "ZenHub" is now required to follow this project, then you should announce it (and provide some instructions, e.g. a link to the relevant resources).

If not, then you should not mention it (to not confuse anyone who tries to follow the project).

> This issue represents the dream

A worthy dream!

But even more: this is a **must** requirement, see below:

# Deployment

## Requirements

### Must

The deployment mechanism must...

* be fully automated
* provide a "dry-run" option (omitting the final publishing-steps)
* provide a "stage-run" option (deploying to a staging-systems)
* provide a "security-fix" option (adopted process for security-related critical fixes)
* be limited to project-leads (e.g. keys needed)
* be implemented with local technologies (e.g. dry-run works off-line)

### Should

* Rely on standard tools and technologies (as much as possible, see "off-line" must-requirement)

## Specification / Implementation

The specification should be provided by a "process fanatic", ideally in parallel with a general project organization.

**Comment by danfinlay** [comment]
> If "ZenHub" is now required to follow this project, then you should announce it (and provide some instructions, e.g. a link to the relevant resources).

We're just experimenting with it, to try to organize our enormous backlog. Not ready for formal "requirement" announcements. We'll add any formal process documentation to our main READMEs as those processes solidify.

**Comment by lazaridiscom** [comment]
> We're just experimenting with it, to try to organize our enormous backlog. 

This will not work-out, as you have to hit the root-causes.

Your project, your free to try, of course.

But please be aware that you make it very difficult for others to follow this project.
